### ISSUE TITLE
Clear error message when compiling an empty MLIR module

### ISSUE THREAD

**Issue by wangkuiyi** [iree-org/iree#12228]
### Request description

If I compile an empty MLIR module like the following
```bash
echo "module @hash_emb { }" | iree-compile -  --iree-hal-target-backends=llvm-cpu
```
I got a confusing error message complaining something about the **outer module**.
```
<stdin>:1:1: error: outer module does not contain a vm.module op
module @hash_emb { }
^
<stdin>:1:1: note: see current operation:
```

Similarly, if I compile an MLIR module containing only variables but no functions, I got the same error message.  
```
$ cat <<EOF | iree-compile -  --iree-hal-target-backends=llvm-cpu
> module @hash_emb {
>   "ml_program.global"() {is_mutable, sym_name = "_x$0", sym_visibility = "private", type = tensor<4x4xf32>, value = dense<1.000000e+00> : tensor<4x4xf32>} : () -> ()
> }
> EOF
<stdin>:1:1: error: outer module does not contain a vm.module op
module @hash_emb {
^
<stdin>:1:1: note: see current operation:
```

**None of the above two cases seem anything with an outer module**.

And, as long as I add a function into the MLIR module, the above error message disappears.
```
cat <<EOF | iree-compile -  --iree-hal-target-backends=llvm-cpu
module @hash_emb {
  "ml_program.global"() {is_mutable, sym_name = "_x$0", sym_visibility = "private", type = tensor<4x4xf32>} : () -> ()
  func.func @run(%arg0: tensor<4x4xf32>) -> tensor<4x4xf32> {
    return %arg0 : tensor<4x4xf32>
  }
}
EOF
```

### What component(s) does this issue relate to?

Compiler

### Additional context

13a4a145b4892f8c146734f5d60afa4ca48ff104

**Comment by GMNGeoffrey** [comment]
Well your module is the outer module :-) And if it had a function the compiler would create vm.module ops, but it doesn't and then some check later on is expecting some to exist. Agreed this should be a better error message though

**Comment by GMNGeoffrey** [comment]
https://github.com/iree-org/iree/blob/6d4aadbe92f89f46211fc19b45bdedfe8d757d36/compiler/src/iree/compiler/Dialect/VM/Target/Bytecode/BytecodeModuleTarget.cpp#L690 is the location of this check (similar one in the c target). It seems like we could just exit with a success result there? Compiling an empty program is likely not super useful, but it seems like something that should work. We could add a warning instead. @benvanik would stuff break if we just returned `success()` there?

**Comment by GMNGeoffrey** [comment]
Hmm no it looks like that function actually is expecting exactly one vm.module perhaps and it calls into something that actually generates the vmfb. I guess we would want an empty vm.module there instead

**Comment by GMNGeoffrey** [comment]
I think I see what's going wrong, and I agree this shouldn't fail this way (or perhaps at all), but what are you actually trying to accomplish by compiling an empty module?

**Comment by wangkuiyi** [comment]
OK. This is the story.

I started with a module that wasn't empty. There are int64-tensor variables, and there are functions that return those variables. But when I try to compile this module, I get the error message below.
```
<stdin>:13:10: error: 'ml_program.global_load' op cannot load from global typed 'tensor<681022xi64>' as 'tensor<681022xi32>'
```

What the message says I did, I think I didn't do. So, I decided to clear out the module and try adding variables and functions back one by one. I thought that IREE would be able to compile the empty module, but that I would get an error when I add some certain variable or function.

It turned out that it doesn't work from the start. :-D

**Comment by wangkuiyi** [comment]
Does it make sense to keep IREE working as is, but just add a few words into the error message

```
<< "outer module does not contain a vm.module op"; 
```

and make it

```
<< "outer module does not contain a vm.module op\n"
<< "Are you trying to bulid an empty MLIR module?"
<< "Do you mind add at least an empty function, so IREE";
<< "could generate a vm.module op in your MLIR module and make it look a little bit more useful?"
```

🤣

**Comment by GMNGeoffrey** [comment]
Yeah I think we should either immediately fail if there are no functions with a clear error message (e.g. compiling a C binary requires a `main`) or compile an empty module. I'm inclined towards the latter, since we allow multiple entry functions and don't have a special "main" and 0 is a whole number :-)

Seems like you're unblocked, but we should still fix this

**Comment by benvanik** [comment]
empty modules are fine - we are not producing executable binaries, we're producing modules, and clang is fine if you give it empty inputs:
```sh
ben@noxa-pc:/mnt/d/Dev/iree$ echo "" | clang -x c -shared -
ben@noxa-pc:/mnt/d/Dev/iree$ file a.out
a.out: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=c130f08078e8499ebc109e24431797bf4bf35f43, not stripped
```

**Comment by GMNGeoffrey** [comment]
Yeah, I agree we should just allow them :-)
